src/foam/mlang/mlang.js

```diff
@@ -3775,6 +3776,137 @@ foam.CLASS({
   ]
 });
 
+foam.CLASS({
+  package: 'foam.mlang',
+  name: 'Formula',
+  extends: 'foam.mlang.AbstractExpr',
+  abstract: true,
+
+  documentation: 'Formula base-class',
+
+  properties: [
+    {
+      class: 'foam.mlang.ExprArrayProperty',
+      name: 'args'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'f',
+      javaCode: `
+        var result = 0.0;
+        for ( int i = 0; i < getArgs().length; i++) {
+          var current = getArgs()[i].f(obj);
+          if ( current instanceof Number ) {
+            var value = ((Number) current).doubleValue();
+            result = result == 0.0 ? value : reduce(result, value);
+          }
+        }
+        return result;
+      `
+    },
+    {
+      name: 'reduce',
+      type: 'Double',
+      abstract: true,
+      args: [
+        { name: 'arg1', type: 'Double' },
+        { name: 'arg2', type: 'Double' }
+      ]
+    }
+  ]
+})
+
+foam.CLASS({
+  package: 'foam.mlang.expr',
+  name: 'Add',
+  extends: 'foam.mlang.Formula',
+  implements: [ 'foam.core.Serializable' ],
+
+  methods: [
+    {
+      name: 'reduce',
+      abstract: false,
+      javaCode: 'return arg1 + arg2;'
+    }
+  ]
+});
+
+foam.CLASS({
+  package: 'foam.mlang.expr',
+  name: 'Subtract',
+  extends: 'foam.mlang.Formula',
+  implements: [ 'foam.core.Serializable' ],
+
+  methods: [
+    {
+      name: 'reduce',
+      abstract: false,
+      javaCode: 'return arg1 - arg2;'
+    }
+  ]
+});
+
+foam.CLASS({
+  package: 'foam.mlang.expr',
+  name: 'Multiply',
+  extends: 'foam.mlang.Formula',
+  implements: [ 'foam.core.Serializable' ],
+
+  methods: [
+    {
+      name: 'reduce',
+      abstract: false,
+      javaCode: 'return arg1 * arg2;'
+    }
+  ]
+});
+
+foam.CLASS({
+  package: 'foam.mlang.expr',
+  name: 'Divide',
+  extends: 'foam.mlang.Formula',
+  implements: [ 'foam.core.Serializable' ],
+
+  methods: [
+    {
+      name: 'reduce',
+      abstract: false,
+      javaCode: 'return arg1 / arg2;'
```

Comment:
can we trap and report division by zero?

Reply:
Addressed in https://github.com/foam-framework/foam2/pull/3628.